### DISCUSSION TITLE
Password reset SMTP "kind: InvalidData,, error: CorruptMessage"

### DISCUSSION THREAD

**Discussion by Entheologian** [lldap/lldap#517] [Q&A]
Hi,

I have these SMTP options set and when I try to reset a password for an existing account using the username, I see the message, but no email is sent and no errors in logs (with verbose enabled):
> A reset token has been sent to your email.

If I try using with the email address, I get this message:
> Could not initiate password reset[500 ]: Internal server error: `Could not send email: Connection error: Connection error: received corrupt message

These are the SMTP options I have set:
```
[smtp_options]
enable_password_reset=true
server="smtp.gmail.com"
port=587
smtp_encryption="TLS"
user="user.name@gmail.com"
password="gmail-app-password"
from="LLDAP Admin <user.name@gmail.com>"
```

The logs show this (obfuscated contact info):
```
lldap_1  | 2023-04-01T11:25:50.952576132+00:00  INFO     HTTP request [ 2.00ms | 4.77% / 100.00% ]
lldap_1  | 2023-04-01T11:25:50.952583880+00:00  INFO     ┝━ ｉ [info]:  | uri: /auth/reset/step1/user.name@gmail.com
lldap_1  | 2023-04-01T11:25:50.952614369+00:00  DEBUG    ┝━ get_password_reset_step1 [ 1.91ms | 47.41% / 95.23% ]
lldap_1  | 2023-04-01T11:25:50.952621555+00:00  DEBUG    │  ┝━ list_users [ 444µs | 22.15% ]
lldap_1  | 2023-04-01T11:25:50.952638499+00:00  DEBUG    │  │  ┝━ 🐛 [debug]:  | filters: Some(Or([UserId(UserId("user.name@gmail.com")), Equality(Email, "user.name@gmail.com")]))
lldap_1  | 2023-04-01T11:25:50.953470556+00:00  DEBUG    │  │  ┕━ 🐛 [debug]:  | return: [UserAndGroups { user: User { user_id: UserId("username"), email: "user.name@gmail.com", display_name: Some("F. Lastname"), first_name: Some("Firstname"), last_name: Some("Lastname"), avatar: Some(JpegPhoto([])), creation_date: 2023-01-26T15:18:34.822650008, uuid: Uuid("0ea98217-8c1f-32d7-9c23-bc16cce20b29") }, groups: None }]                                                             
lldap_1  | 2023-04-01T11:25:50.953482935+00:00  DEBUG    │  ┝━ start_password_reset [ 514µs | 25.67% ]
lldap_1  | 2023-04-01T11:25:50.953488432+00:00  DEBUG    │  │  ┕━ 🐛 [debug]:  | user: UserId("username")
lldap_1  | 2023-04-01T11:25:50.956719808+00:00  DEBUG    │  ┝━ 🐛 [debug]: Sending email to 'user.name@gmail.com' as 'LLDAP Admin <user.name@gmail.com>' via 'user.name@gmail.com'@'smtp.gmail.com':'587'
lldap_1  | 2023-04-01T11:25:50.986287536+00:00  DEBUG    │  ┝━ 🐛 [debug]: No cached session for DnsName(DnsName(DnsName("smtp.gmail.com"))) | log.target: "rustls::client::hs" | log.module_path: "rustls::client::hs" | log.file: "/__w/lldap/lldap/${GITHUB_WORKSPACE}/.cargo/registry/src/github.com-1ecc6299db9ec823/rustls-0.20.8/src/client/hs.rs" | log.line: 55
lldap_1  | 2023-04-01T11:25:50.986382841+00:00  DEBUG    │  ┝━ 🐛 [debug]: Not resuming any session | log.target: "rustls::client::hs" | log.module_path: "rustls::client::hs" | log.file: "/__w/lldap/lldap/${GITHUB_WORKSPACE}/.cargo/registry/src/github.com-1ecc6299db9ec823/rustls-0.20.8/src/client/hs.rs" | log.line: 127
lldap_1  | 2023-04-01T11:25:51.020981309+00:00  WARN     │  ┕━ 🚧 [warn]: Error sending email: lettre::transport::smtp::Error {
lldap_1  |     kind: Connection,
lldap_1  |     source: lettre::transport::smtp::Error {
lldap_1  |         kind: Connection,
lldap_1  |         source: Custom {
lldap_1  |             kind: InvalidData,
lldap_1  |             error: CorruptMessage,
lldap_1  |         },
lldap_1  |     },
lldap_1  | }
lldap_1  | 2023-04-01T11:25:51.021014089+00:00  INFO     ┕━ ｉ [info]:  | status_code: 500
```

Did I mis-configure the email settings?

**Comment by Karstensson** [comment]
I have the exact same issue :-(

**Reply by nitnelave** [reply]
Hi! Receiving a `CorruptMessage` means that the SMTP encryption setting is misconfigured: between TLS, StartTls and None, your SMTP server expects one on the port you gave but you send another.
I think that for gmail, the vocabulary is in a sad state: TLS means StartTls, and SSL means TLS. Try changing the port to 465? Or the encryption method to StartTls (but it's safer to use the SSL/465 port)

**Reply by Karstensson** [reply]
Wow, thanks for a quick reply! There is no way to select SSL in lldap_config, only options are None, TLS and START TLS.

When I runt TLS on port 465 I get:
`2023-10-06T22:35:28.831231581+00:00  WARN     │  ┝━ 🚧 [warn]: Error sending email: lettre::transport::smtp::Error {
    kind: Permanent(
        Code {
            severity: PermanentNegativeCompletion,
            category: Unspecified3,
            detail: Five,
        },
    ),
    source: "5.7.8 Username and Password not accepted. Learn more at5.7.8  https://support.google.com/mail/?p=BadCredentials i25-20020a1709064ed900b009b947aacb4bsm3489498ejv.191 - gsmtp",
}`

When I run STARTTLS port port 465 I get
`2023-10-06T22:36:42.562941533+00:00  DEBUG    │  ┝━ 🐛 [debug]: Sending email to user@gmail.com' as 'LLDAP Admin <serverEmail@gmail.com>' via 'serverEmail@gmail.com'@'smtp.gmail.com':'465'
2023-10-06T22:36:52.672136425+00:00  WARN     │  ┝━ 🚧 [warn]: Error sending email: lettre::transport::smtp::Error {
    kind: Response,
    source: "incomplete response",
}
2023-10-06T22:36:52.672140836+00:00  INFO     │  ┕━ ｉ [info]: Reset token: r8ZBmKfIwlKps8DyHF84slx2Vl8M3WyLHhKdbhG85LW2n07xVAHksqDx86RULgVCD096XFd5mHCrHyY1mchSHXfnaIszvWQyOk2q
2023-10-06T22:36:52.672168105+00:00  DEBUG    ┕━ 🐛 [debug]:  | status_code: 500`

When I run STARTLS on påort 587 i get
`2023-10-06T22:38:54.919918504+00:00  WARN     │  ┝━ 🚧 [warn]: Error sending email: lettre::transport::smtp::Error {
    kind: Permanent(
        Code {
            severity: PermanentNegativeCompletion,
            category: Unspecified3,
            detail: Five,
        },
    ),
    source: "5.7.8 Username and Password not accepted. Learn more at5.7.8  https://support.google.com/mail/?p=BadCredentials by8-20020a170906a2c800b009b913aa7cdasm3486810ejb.92 - gsmtp",
}`

**Comment by nitnelave** [comment]
Right, when I was referring to SSL it's because it's what you find when you
look up Google's SMTP settings. In our case, I'd recommend TLS with port
465. The response you got (invalid credentials) is from Google, so it's
working as intended. You probably need an app password to be able to
connect with just a password to Gmail, though.

**Reply by Karstensson** [reply]
Hmm, then I'll have to investigate this further. I am already using an app password. I also have smtp with gmail up and running with authelia, mealie and vaultwarden, but then I use 587 with TLS. I also tried to renew the app password and also tried to use the password I have from mealie, but that didn't seem to work either.

**Reply by Karstensson** [reply]
I feel embarrassed, but I managed to get it working. Exactly as you and the error message mentioned it was an authentication error. There was a bad character in the username I was using. So end settings to make it work was.

port=465
smtp_encryption = "TLS"

So once again, thank you very much for your help!

**Reply by nitnelave** [reply]
Great! I'm glad you found the typo :)

Once again, I'd recommend SSL and port 587, it's just a safer standard. But both work
